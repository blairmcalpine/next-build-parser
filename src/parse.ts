#!/usr/bin/env node

import { findMarkers } from './sanitize.js';
import {
  generateMiddleware,
  generateRoutes,
  generateRouteTypeMap,
  generateSharedByAll,
  generateSharedByAllTotal,
} from './route.js';
import { defaultSizeUnit, type SizeUnit } from './unit.js';

export type ParseOptions = {
  unit?: SizeUnit;
};

export const parse = (nextBuildOutput: string, options?: ParseOptions) => {
  const data = nextBuildOutput.split('\n').filter(Boolean);
  const unit = options?.unit ?? defaultSizeUnit;

  const {
    sharedByAllMarker,
    routeTypeMapMarker,
    routeMarker,
    middlewareMarker,
  } = findMarkers(data);

  const routeRows = data.slice(routeMarker + 1, sharedByAllMarker);
  const sharedByAllRows = data.slice(
    sharedByAllMarker + 1,
    middlewareMarker ?? routeTypeMapMarker,
  );
  const routeTypeMapRows = data.slice(routeTypeMapMarker);
  const sharedByAllTotalRow = data[sharedByAllMarker]!;
  const middlewareRow =
    middlewareMarker !== undefined ? data[middlewareMarker] : undefined;

  const routeTypeMap = generateRouteTypeMap(routeTypeMapRows);
  const routes = generateRoutes(routeRows, routeTypeMap, unit);
  const shared = generateSharedByAll(sharedByAllRows, unit);
  const sharedTotal = generateSharedByAllTotal(sharedByAllTotalRow, unit);
  const middleware = generateMiddleware(middlewareRow, unit);

  const output = {
    ...(middleware !== undefined ? { middleware } : {}),
    routes,
    shared,
    sharedTotal,
    unit,
  };
  return output;
};
