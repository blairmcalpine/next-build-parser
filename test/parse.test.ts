import { expect, describe, it } from 'vitest';
import { parse } from '../src/parse.js';
import fs from 'node:fs';

describe('parse', () => {
  describe('routes', () => {
    it('should output based on the unit option', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { routes } = parse(file, {
        unit: 'B',
      });
      const routeSizes = routes.map((route) => route.size);
      const routeFirstLoads = routes.map((route) => route.firstLoad);
      expect(routeSizes).toMatchSnapshot();
      expect(routeFirstLoads).toMatchSnapshot();
    });

    it('should output the correct route type', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { routes } = parse(file);
      const routeTypes = routes.map((route) => route.type);
      expect(routeTypes).toMatchSnapshot();
    });

    it('should output the correct route icon', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { routes } = parse(file);
      const routeIcons = routes.map((route) => route.icon);
      expect(routeIcons).toMatchSnapshot();
    });

    it('should output the correct route name', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { routes } = parse(file);
      const routePaths = routes.map((route) => route.name);
      expect(routePaths).toMatchSnapshot();
    });
  });

  describe('shared', () => {
    it('should output based on the unit option', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { shared } = parse(file, {
        unit: 'B',
      });
      expect(shared).toMatchSnapshot();
    });
    it('should output the correct names', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { shared } = parse(file);
      const sharedNames = shared.map((shared) => shared.name);
      expect(sharedNames).toMatchSnapshot();
    });
    it('should output the correct shared by all total', () => {
      const file = fs.readFileSync('test/samples/basic.txt', 'utf8');
      const { sharedTotal } = parse(file);
      expect(sharedTotal).toMatchSnapshot();
    });
    it('should output the correct other chunks', () => {
      const file = fs.readFileSync('test/samples/other-shared.txt', 'utf8');
      const { shared } = parse(file);
      const otherChunks = shared.filter((shared) => shared.name === 'other');
      expect(otherChunks).toMatchSnapshot();
    });
  });
  describe('middleware', () => {
    it('should output based on the unit option', () => {
      const file = fs.readFileSync('test/samples/middleware.txt', 'utf8');
      const { middleware } = parse(file, {
        unit: 'B',
      });
      expect(middleware).toMatchSnapshot();
    });
  });
});
