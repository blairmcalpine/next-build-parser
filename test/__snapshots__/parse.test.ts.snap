// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > middleware > should output based on the unit option 1`] = `0`;

exports[`parse > routes > should output based on the unit option 1`] = `
[
  1126.4,
  875,
  0,
  15872,
  1126.4,
  8345.6,
  0,
  1126.4,
]
`;

exports[`parse > routes > should output based on the unit option 2`] = `
[
  128000,
  88268.8,
  0,
  143360,
  128000,
  135168,
  0,
  128000,
]
`;

exports[`parse > routes > should output the correct route icon 1`] = `
[
  "λ",
  "λ",
  "λ",
  "λ",
  "λ",
  "λ",
  "ℇ",
  "λ",
]
`;

exports[`parse > routes > should output the correct route name 1`] = `
[
  "/",
  "/_not-found",
  "/api/trpc/[trpc]",
  "/create",
  "/popular",
  "/puzzle/[id]",
  "/puzzle/[id]/opengraph-image",
  "/recent",
]
`;

exports[`parse > routes > should output the correct route type 1`] = `
[
  "Dynamic",
  "Dynamic",
  "Dynamic",
  "Dynamic",
  "Dynamic",
  "Dynamic",
  "Edge",
  "Dynamic",
]
`;

exports[`parse > shared > should output based on the unit option 1`] = `
[
  {
    "name": "chunks/472-9100b5fcfec8f88c.js",
    "size": 30720,
  },
  {
    "name": "chunks/fd9d1056-853251486b1823c6.js",
    "size": 54579.2,
  },
  {
    "name": "chunks/main-app-842b2bfec3a562ae.js",
    "size": 231,
  },
  {
    "name": "chunks/webpack-b7a1470890cdd00b.js",
    "size": 1832.96,
  },
]
`;

exports[`parse > shared > should output the correct names 1`] = `
[
  "chunks/472-9100b5fcfec8f88c.js",
  "chunks/fd9d1056-853251486b1823c6.js",
  "chunks/main-app-842b2bfec3a562ae.js",
  "chunks/webpack-b7a1470890cdd00b.js",
]
`;

exports[`parse > shared > should output the correct other chunks 1`] = `[]`;

exports[`parse > shared > should output the correct shared by all total 1`] = `85.3`;
